### ISSUE TITLE
Create share commitments using nmt instead of a normal merkle tree

### ISSUE THREAD

**Issue by evan-forbes** [celestiaorg/celestia-app#1004]
Currently, we are creating the share commitments by taking a normal merkle root over the nmt subtree roots. We should use an nmt over the subtree roots as well.

**Comment by nashqueue** [comment]
We thought about doing this for the row and column roots as well. The problem here is that the nodes in a nmt need to be ordered by nid . The nid of a parity share is 0xfffff.... . This means when you erasure code a column, (c1/r2k) will have the nid of 0xfffff.... and (c2/r1) will have a nid of most likely 0. But in the nmt over the column roots (c1/r2k) will be on the left of (c2/r1). This breaks the nmt making it not possible to construct a nmt over the col & row roots as we are currently constructing them.

**Comment by nashqueue** [comment]
One question that came up if the nmt is required to be a binary tree. This might not be the case as the nmt is forked from nebulous which does not require a binary tree.

**Comment by adlerjohn** [comment]
There is an assumption in the NMT algorithm that the tree is perfect (i.e. the leaves are a power of 2).

**Comment by adlerjohn** [comment]
It's not clear to me what the benefit/rationale of using an NMT is. All shares have the same known namespace, so there's no need to know the min/max NID of the shares. The commitment won't be identical above the subtree roots for blobs that span multiple rows since parity shares aren't accounted for in the PFB commitment.

**Comment by liamsi** [comment]
> There is an assumption in the NMT algorithm that the tree is perfect (i.e. the leaves are a power of 2).

Not advocating for anything here but just noting that is sth we introduced and could change in NMT.

**Comment by evan-forbes** [comment]
yes I think we should do this if we can fit it in after thinking about it. The main benefit would be that other implementations would be able to use a single merkle tree implementation.

**Comment by nashqueue** [comment]
There is a consideration that we could improve ZK-rollup computation by not having a tree but only one hash. The question is if there is a situation where we need a Merkle proof of the commitment from the subtree roots. Then we would increase this inclusion proof by having all subtree roots as siblings.

**Comment by rootulp** [comment]
Just noting that there is a size downside to using an NMT over the subtree roots:

Tree | Commitment size
--- | ---
Binary Merkle Tree | 32 bytes
Namespaced Merkle Tree | 98 bytes

**Comment by nashqueue** [comment]
Let's say we live in a world where we get our commitment proofs by PFB Transactions, as proposed in ADR 11. Now that a fraud poof happens in the state machine of rollkit we would normally need a Merkle inclusion proof to the shares in which the fraud exists for compact fraud proofs. The question is whether we could get smaller inclusion proofs of the Merkle Tree commitment rather than the RowRoots. And in some rare cases, this could be true. In most cases, this will not be true. I don't see a merkle inclusion proof to the commitment being useful.
